From 02a5acc50a462f1d38277e8299dbfaa8fdf8a21a Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Wed, 10 Apr 2024 15:21:29 +0800
Subject: [PATCH 60/75] MMFMWK-9360 [imx95]v4l2: return OK instead of EOS when
 sending gap event

For below  2 streams, there is no IDR frames queued after seek to 20%,
normal I and P frames before next IDR after seek will be skipped. Decoder
will send gap event if over 50 output frames cannot be decoded. So need
to return GST_FLOW_OK instead of GST_FLOW_EOS limited to one multiqueue
commit 77919cbf33ea2eaa854a8764e9c5deaed35ae4ce on gst1.24.

/mnt/streams/ttVector/SHAVectors/H264Dec/Conformance/BP/H264_BP20_352x288_24_1887_MP3_24_56_2_ray.avi
/mnt/streams/ttVector/SHAVectors/H264Dec/Conformance/MP/H264_MP21_704x352_30_305_MP3_48_69_1_vbr.avi

Upstream-Status: Pending
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 sys/v4l2/gstv4l2allocator.c |  6 +++---
 sys/v4l2/gstv4l2object.c    |  1 -
 sys/v4l2/gstv4l2object.h    |  1 -
 sys/v4l2/gstv4l2videodec.c  | 22 +++++++++++-----------
 4 files changed, 14 insertions(+), 16 deletions(-)

diff --git a/sys/v4l2/gstv4l2allocator.c b/sys/v4l2/gstv4l2allocator.c
index c4058f9..15d7bb9 100644
--- a/sys/v4l2/gstv4l2allocator.c
+++ b/sys/v4l2/gstv4l2allocator.c
@@ -1435,13 +1435,13 @@ gst_v4l2_allocator_dqbuf (GstV4l2Allocator * allocator,
   if (V4L2_TYPE_IS_OUTPUT (obj->type)) {
     gst_v4l2_allocator_reset_size (allocator, group);
     if (group->buffer.flags & V4L2_BUF_FLAG_ERROR) {    //output buffer cannot be decoded
+      GST_INFO_OBJECT (allocator, "OUTPUT error flag for frame %u",
+          (guint32) group->buffer.timestamp.tv_sec);
       obj->drop_frames =
           g_list_append (obj->drop_frames, &group->buffer.timestamp);
-      if (!obj->frame_decoded)
-        obj->err_cnt++;
+      obj->err_cnt++;
     } else {
       obj->err_cnt = 0;
-      obj->frame_decoded = TRUE;
     }
   } else {
     if (group->buffer.flags & V4L2_BUF_FLAG_ERROR) {    //capture buffer has error flag
diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 6508969..3adffb5 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -571,7 +571,6 @@ gst_v4l2_object_new (GstElement * element,
 
   v4l2object->drm_modifier = 0;
   v4l2object->is_g2 = FALSE;
-  v4l2object->frame_decoded = FALSE;
   v4l2object->err_cnt = 0;
 
   v4l2object->keep_aspect = TRUE;
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index 1c9204f..82ed45b 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -156,7 +156,6 @@ struct _GstV4l2Object {
   gboolean has_alpha_component;
   guint64 drm_modifier;
   gboolean is_g2;
-  gboolean frame_decoded;
   guint32 err_cnt;
 
   /* only used if the device supports MPLANE
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index fb3ad86..7e9136f 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -358,7 +358,6 @@ gst_v4l2_video_dec_flush (GstVideoDecoder * decoder)
     return TRUE;
 
   self->output_flow = GST_FLOW_OK;
-  self->v4l2output->frame_decoded = FALSE;
   self->v4l2output->err_cnt = 0;
 
   gst_v4l2_object_unlock_stop (self->v4l2output);
@@ -1260,17 +1259,17 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
     }
     GST_VIDEO_DECODER_STREAM_LOCK (decoder);
 
+    /* If there are over 50 continuous frames that cannot be decoded from start
+      * or after seek, send gap event to finish preroll */
     if (self->v4l2output->err_cnt > MAX_OUTPUT_ERROR_COUNT) {
-      if (!self->v4l2output->frame_decoded) {
-        self->v4l2output->frame_decoded = TRUE;
-
-        /* If there are over 50 continuous frames that cannot be decoded from start
-         * or after seek, send gap event to finish preroll */
-        gst_pad_push_event (decoder->srcpad, gst_event_new_gap (0,
-                GST_CLOCK_TIME_NONE));
-      }
-      ret = GST_FLOW_EOS;
-      goto drop;
+      gst_pad_push_event (decoder->srcpad, gst_event_new_gap (0,
+              GST_CLOCK_TIME_NONE));
+      ret = GST_FLOW_OK;
+
+      /* If OUTPUT buffer is dequeued immediately after queuing, it means this frame
+       * has been dropped, so just unref the pool and return */
+      if (!frame)
+        goto out;
     }
 
     if (ret == GST_FLOW_FLUSHING) {
@@ -1292,6 +1291,7 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
   gst_buffer_unref (tmp);
 
   gst_video_codec_frame_unref (frame);
+out:
   if (pool)
     gst_object_unref (pool);
   return ret;
-- 
2.43.0

